Metadata/DynamicListMetadataLoader.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+
+namespace Sulu\Bundle\FormBundle\Metadata;
+
+use Sulu\Bundle\AdminBundle\Metadata\ListMetadata\FieldMetadata;
+use Sulu\Bundle\AdminBundle\Metadata\ListMetadata\ListMetadata;
+use Sulu\Bundle\AdminBundle\Metadata\ListMetadata\ListMetadataLoaderInterface;
+use Sulu\Bundle\FormBundle\Dynamic\FormFieldTypePool;
+use Sulu\Bundle\FormBundle\Entity\Form;
+use Sulu\Bundle\FormBundle\ListBuilder\DynamicListFactory;
+use Sulu\Bundle\FormBundle\Manager\FormManager;
+use Symfony\Component\Translation\TranslatorInterface;
+
+class DynamicListMetadataLoader implements ListMetadataLoaderInterface
+{
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @var FormManager
+     */
+    private $formManager;
+
+    /**
+     * @var DynamicListFactory
+     */
+    private $dynamicListFactory;
+
+    public function __construct(
+        TranslatorInterface $translator,
+        FormManager $formManager,
+        DynamicListFactory $dynamicListFactory
+    ) {
+        $this->translator = $translator;
+        $this->formManager = $formManager;
+        $this->dynamicListFactory = $dynamicListFactory;
+    }
+
+
+    public function getMetadata(string $key, string $locale, array $metadataOptions): ?ListMetadata
+    {
+        if (strcmp('form_data', $key) !== 0) {
+            return null;
+        }
+
+        $list = new ListMetadata();
+
+        $form = $this->getForm($metadataOptions, $locale);
+        if (!$form) {
+            return null;
+        }
+
+        $fieldDescriptors = $this->dynamicListFactory->getFieldDescriptors($form, $locale);
+        foreach ($fieldDescriptors as $fieldDescriptor) {
+            $field = new FieldMetadata($fieldDescriptor->getName());
+            $field->setLabel($this->translator->trans($fieldDescriptor->getTranslation(), [], 'admin', $locale));
+            $field->setType($fieldDescriptor->getType());
+            $field->setVisibility($fieldDescriptor->getVisibility());
+            $field->setSortable($fieldDescriptor->getSortable());
+
+            $list->addField($field);
+        }
+
+        $list->setCacheable(false);
+        return $list;
+    }
+
+    private function getForm(array $metadataOptions, string $locale): ?Form
+    {
+        if (!array_key_exists('id', $metadataOptions)) {
+            return null;
+        }
+        $entity = $this->formManager->findById($metadataOptions['id'], $locale);
+
+        if (!$entity) {
```

Comment:
Isn't that already null when `findById` doesn't find anything? Than this `if` could be omitted.